odl/space/cupy_tensors.py

```diff
@@ -0,0 +1,2248 @@
+
+        Parameters
+        ----------
+        inp : `array-like`, optional
+            Input used to initialize the new element.
+
+            If ``inp`` is `None`, an empty element is created with no
+            guarantee of its state (memory allocation only).
+            The new element will use ``order`` as storage order if
+            provided, otherwise `default_order`.
+
+            Otherwise, a copy is avoided whenever possible. This requires
+            correct `shape` and `dtype`, and if ``order`` is provided,
+            also contiguousness in that ordering. If any of these
+            conditions is not met, a copy is made.
+
+        order : {None, 'C', 'F'}, optional
+            Storage order of the returned element. For ``'C'`` and ``'F'``,
+            contiguous memory in the respective ordering is enforced.
+            The default ``None`` enforces no contiguousness.
+
+        Returns
+        -------
+        element : `CupyTensor`
+            The new element created (from ``inp``).
+
+        Notes
+        -----
+        This method preserves "array views" of correct size and type,
+        see the examples below.
+
+        Examples
+        --------
+        >>> space = odl.rn((2, 3), impl='cupy')
+
+        Create an empty element:
+
+        >>> empty = space.element()
+        >>> empty.shape
+        (2, 3)
+
+        Initialization during creation:
+
+        >>> x = space.element([[1, 2, 3],
+        ...                    [4, 5, 6]])
+        >>> x
+        rn((2, 3), impl='cupy').element(
+            [[ 1.,  2.,  3.],
+             [ 4.,  5.,  6.]]
+        )
+        """
+        if order is None:
+            order_in = order
+        else:
+            order, order_in = str(order).upper(), order
+
+        if order is not None and order not in ('C', 'F'):
+            raise ValueError("`order` {!r} not understood".format(order_in))
+
+        with cupy.cuda.Device(self.device):
+            if inp is None:
+                if order is None:
+                    order = self.default_order
+                arr = cupy.empty(self.shape, dtype=self.dtype, order=order)
+
+            else:
+                if inp in self and order is None:
+                    # Short-circuit for space elements and no enforced ordering
+                    return inp
+
+                if hasattr(inp, 'shape') and inp.shape != self.shape:
+                    raise ValueError('`inp` must have shape {}, got shape {}'
+                                     ''.format(self.shape, inp.shape))
+
+                if isinstance(inp, cupy.ndarray):
+                    # Workaround for https://github.com/cupy/cupy/issues/590
+                    # TODO: remove when solved
+                    if (inp.dtype == self.dtype and
+                            inp.device.id == self.device):
+                        arr = inp
+                    else:
+                        arr = inp.astype(self.dtype)
```

Comment:
What if the device is wrong?

Reply:
No idea 😇 I'll check

Reply:
Or rather: I don't have a way to check right now, so I'll have to check the documentation. I *guess* what happens is that the `.astype()` method creates a copy, and the `cupy.cuda.Device()` context ensures that the new array is created on the given device. But I'll make sure that is true.

Reply:
You can use jupyterhub on radon which has 8 cores to check this

Reply:
Good call!